jax/interpreters/pxla.py

```diff
@@ -579,9 +648,15 @@ def dynamic_fun(dummy, *args):
   compiled = built.Compile(compile_options=compile_options,
                            backend=xb.get_backend(backend))
 
-  handle_args = partial(shard_args, compiled.local_devices(),
-                        assign_shards_to_replicas(num_local_replicas, axis_size),
-                        axis_size)
+  sharding_specs = [_pmap_sharding_spec(num_local_replicas, axis_size,
+                                        aval.shape)
+                    if aval is not core.abstract_unit else None
+                    for aval in sharded_avals]
+  indices = [spec_to_indices(aval.shape, spec)
+             if spec is not None else None
+             for aval, spec in zip(avals, sharding_specs)]
+  handle_args = partial(shard_args, compiled.local_devices(), indices)
+
   handle_outs = _pvals_to_results_handler(axis_size, num_local_replicas,
```

Comment:
Since we've already computed `sharding_specs` and `indices`, shouldn't `_pvals_to_results_handler` take those rather than rebuilding them?

Reply:
`sharding_specs` and `indices` are only for the inputs. The outputs may have different shapes (and different shardings, once things like replication and sharded_jit land). I'll change the variable names to `input_sharding_specs` and `input_indices` as a hint. (I always forget about this too btw.)